### PR TITLE
Replaced event queue based watching resources in router with shared informers

### PR DESCRIPTION
**Pull request by pravisankar**
- Custom shared informer is used to leverage namespace, label and field filtering.
  (Auto generated shared informer does not allow this)
- Listing resources by shared informers doesn't order by resource version/creation time.
  So custom lister for routes is used to order the route list by creation time and this
  will allow oldest route to be processed before new route to claim the host name.
- Synchronization with the informer queue and cache is a bit difficult as the cache could
  have newer changes than what was pushed on to the queue. Luckily We only care about the
  first sync to avoid 503 status code for routes.
- Handling first sync:
  * Informers are started with no registered event handlers
  * Wait for all informers to be synced
  * Block router reload
  * Get list of items from informers store and process manually
  * Perform router reload
  * Register router event handlers
  This guarantees first router sync is performed after processing all existing items.
- Subsequent router syncs rely on informer syncing sate and uses rate limiter to coalesce changes.

- Deleted eventQueue, no longer used

Trello card: https://trello.com/c/y6SFvOA7

### CODE REVIEW

**Comment by pravisankar** [comment]
[test]

**Comment by pravisankar** [comment]
@openshift/networking @knobunc @rajatchopra PTAL

**Comment by knobunc** [comment]
/lgtm Thanks @pravisankar.

@smarterclayton does this look sane to you?  Obviously we still need to look into some of the test case failures (especially the one around the router reload).

**Comment by smarterclayton** [comment]
I don't like the live calls.  Let's do this the proper way and use the cache correctly.  Wait for sync, then flip a bool and force a refresh on the cache.

Also, don't use the resource name switch, just embed five informer inits.  Strong typing is better

**Comment by smarterclayton** [comment]
Note that I'm really glad the event queue is gone, just want to get the last extra mile to make this "normal".  Live calls are bad because they won't take advantage of API chunking when we turn that on in 3.8

**Comment by smarterclayton** [comment]
@deads2k re: how to safely have "only do this after sync" with the informer

I think using an index on hostname and then doing a sort is the right thing
to do, and if the "ready boolean" is unset simply exit the loop and come
back around.  Or we can just delay the initial sync step of writing out the
config until sync is safe.

I do think we don't want to write back to the route API until we've fully
synced, so we probably are going to have to:

1. complete the full sync and populate the index (writing nothing to route
api for the hostname overlap)
2. wait for that
3. trigger a refresh that will then do the exact same work over again but
will trigger hostname writes to route api
4. let the first write happen

**Comment by deads2k** [comment]
> @deads2k re: how to safely have "only do this after sync" with the informer

In general, "do this after sync" is expressed by filling a work queue while the cache is priming, but not starting any workers.  After all caches have sync, you can synchronously do some work (fill a secondary cache perhaps?) and then start a single worker.

Doing it like that would ensure let you trigger based off of a shared informer, do some work before you consume and process any update, and process resources in order. If you must process individual watch notifications (not resources), then you could fill your own queue (or super deep channel).

**Comment by smarterclayton** [comment]
If you need additional feedback this week to get this closed out please
don't hesitate to ask.  I'd like this in 3.7.

**Comment by pravisankar** [comment]
[test]

**Comment by pravisankar** [comment]
@smarterclayton @knobunc @rajatchopra Updated, PTAL

File: pkg/router/controller/factory/factory.go [hunk 1]
```diff
@@ -2,509 +2,265 @@ package factory
...
-		field = lw.field.String()
-	}
-	opts := metav1.ListOptions{
-		LabelSelector: label,
-		FieldSelector: field,
-	}
-	routes, err := lw.client.Routes(lw.namespace).List(opts)
-	if err != nil {
-		return nil, err
+	rc.FirstSyncDone = func() bool {
```
Comment (smarterclayton) [inline pkg/router/controller/factory/factory.go]: Once has synced has returned true, this is unnecessary.

Maybe we should talk in person, I still think this is much more complicated than it needs to be.

1. Set "syncing" boolean true (never commit while this is true)
2. have your config loops update internal structs
3. wait for all informers synced
4. set syncing to be true
5. trigger a refresh in informers

3-5 should be able to be done in a single method.

**Comment by smarterclayton** [comment]
Regarding the question today, if an informer index of routes by host is
used then the index is up to date once synced is complete (index updates
are synchronous).  So we can replace our internal map with the index, and
if we set the Boolean after true and resync any handlers are guaranteed to
observe the other routes with the same host.

Alternatively, we can simply avoid registering our handlers until sync is
true, and then register and resync (may have to double check the ordering
guarantees around adding a handler).


On Sep 26, 2017, at 9:06 PM, Ravi Sankar Penta <notifications@github.com>
wrote:

/test extended_conformance_gce

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/16315#issuecomment-332378835>,
or mute
the thread

**Comment by smarterclayton** [comment]
Because this is so critical it can be delivered post dcut, although i don't
want to slip too much.

File: pkg/router/controller/router_controller.go [hunk 2]
```diff
@@ -183,56 +105,36 @@ func (c *RouterController) HandleNode() {
...
 
 	if err := c.Plugin.HandleEndpoints(eventType, endpoints); err != nil {
 		utilruntime.HandleError(err)
 	}
-
-	// Change the local sync state within the lock to ensure that all
-	// event handlers have the same view of sync state.
-	c.endpointsListConsumed = c.EndpointsListConsumed()
-	c.commit()
+	c.updateConsumedCount(endpoints)
```
Comment (smarterclayton) [inline pkg/router/controller/router_controller.go]: Counting isn't enough to tell you whether you've got everything.  If an endpoint is deleted while you're doing the sync you will never reach your target number.  You can't synchronize with your handlers like this unfortunately.

File: pkg/router/controller/factory/factory.go (hunk 1 above)
Comment (smarterclayton) [reply pkg/router/controller/factory/factory.go]: So the issue here is that you have two caches:

1. in the informer
2. caches in the route plugin

The second cache is filled by the first cache.  The code that fills the second cache is not done under the lock that maintains the first cache, and therefore while order of events can be guaranteed, there is no guarantee that the second cache is up to date with the first cache.

You need to determine when the second cache has observed all of the events from the first cache once it has synced once.  

I *think* you can address this by starting the informers, waiting for synced (on all of them) and then call each cache as informer.GetStore().List() and send those to the route controller.  Then call commit.  Then register the route controller, and every update from then on is safe.

**Comment by pravisankar** [comment]
@smarterclayton @knobunc Updated, PTAL

File: pkg/router/controller/router_controller.go [hunk 3]
```diff
@@ -72,64 +59,9 @@ func (c *RouterController) Run() {
...
-
-// watchForFirstSync loops until the first sync has been handled.
-func (c *RouterController) watchForFirstSync() {
-	for {
-		if c.handleFirstSync() {
-			return
-		}
-		time.Sleep(50 * time.Millisecond)
-	}
+	c.StartInformers(utilwait.NeverStop)
```
Comment (smarterclayton) [inline pkg/router/controller/router_controller.go]: This doesn't seem to belong here, but in the factory.  Why should this be a concern of the route controller?

Registering functions into another type is a code smell.  Just have a higher level method that calls public methods on the controller from the factory

File: pkg/router/controller/router_controller.go [hunk 4]
```diff
@@ -321,3 +194,61 @@ func (c *RouterController) processIngressEvents(events []ingressRouteEvents) {
 		}
 	}
 }
+
+func (c *RouterController) handleFirstSync() {
+	if err := c.WaitForInformerCacheSync(); err != nil {
+		glog.Fatalf("Failed to sync router informer cache: %v", err)
+	}
+	c.processExistingItems()
+	c.firstSyncDone = true
```
Comment (smarterclayton) [inline pkg/router/controller/router_controller.go]: The only part of this method that really belongs here is setting this Boolean (which needs to be under a lock).

File: pkg/router/controller/router_controller.go [hunk 5]
```diff
@@ -321,3 +194,61 @@ func (c *RouterController) processIngressEvents(events []ingressRouteEvents) {
...
+	glog.V(4).Infof("Router first sync complete")
+	c.Commit()
+}
+
+func (c *RouterController) processExistingItems() {
+	for _, item := range c.InformerCacheList(&kapi.Endpoints{}) {
+		c.HandleEndpoints(watch.Added, item.(*kapi.Endpoints))
+	}
+
+	for _, item := range c.InformerCacheList(&routeapi.Route{}) {
```
Comment (smarterclayton) [inline pkg/router/controller/router_controller.go]: This abstraction is unnecessary.  Do this from the factory and just call the store directly.

**Comment by pravisankar** [comment]
@smarterclayton @knobunc rearranged the code as suggested, please review

File: pkg/router/controller/factory/factory.go [hunk 6]
```diff
@@ -2,509 +2,295 @@ package factory
...
 
-// routeLW is a ListWatcher for routes that can be filtered to a label, field, or
-// namespace.
-type routeLW struct {
-	client    routeclient.RoutesGetter
-	label     labels.Selector
-	field     fields.Selector
-	namespace string
+func (f *RouterControllerFactory) initCallbacks(rc *routercontroller.RouterController) {
+	rc.HasSyncedInformers = func() bool {
```
Comment (smarterclayton) [inline pkg/router/controller/factory/factory.go]: Do you still need this?

File: pkg/router/controller/factory/factory.go [hunk 7]
```diff
@@ -2,509 +2,295 @@ package factory
...
 }
 
-func (lw *routeLW) Watch(options metav1.ListOptions) (watch.Interface, error) {
-	var label, field string
-	if lw.label != nil {
-		label = lw.label.String()
-	}
-	if lw.field != nil {
-		field = lw.field.String()
+func (f *RouterControllerFactory) processExistingItems(rc *routercontroller.RouterController) {
```
Comment (smarterclayton) [inline pkg/router/controller/factory/factory.go]: godoc on this function and how it is used (and reason)

File: pkg/router/controller/factory/factory.go (hunk 6 above)
Comment (pravisankar) [reply pkg/router/controller/factory/factory.go]: When re-sync is in progress, we want to reduce the number of reloads. We could fully rely on router coalescing and can get rid of this informer synced check. @knobunc @rajatchopra what do you think?

Comment (pravisankar) [reply pkg/router/controller/factory/factory.go]: Tested with 1000 routes to check whether router coalescing is sufficient or informer synced check is necessary to reduce reloads. Router coalescing without sync check worked fine, so removed this unnecessary check.

**Comment by pravisankar** [comment]
@smarterclayton @knobunc @rajatchopra can you please take another look?

**Comment by smarterclayton** [comment]
Looks phenomenal, thanks for cleaning up.  Very easy to read now.

/lgtm

**Comment by smarterclayton** [comment]
Was preapproved due to importance to getting this fixed

**Pull request merged**